livecheck: add RuboCops

nandahkrishna opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This PR creates an audit and adds RuboCops for livecheck.

## Audit
Check whether livecheck is able to find versions (doesn't error out) for formulae with livecheck blocks defined.

## RuboCops
1. Ensure that skipped formulae do not have additional livecheck information such as `url`, `regex`, etc.
2. Ensure that a `url` is provided.
3. Ensure that a formula URL is referenced by a symbol (`:head`, `:stable`, `:homepage`).
4. Ensure that `regex`es contain `\.t` and not `\.tgz` or `\.tar\.gz`, etc.
5. Ensure that a `regex` is provided for `strategy :page_match`.
6. Ensure that `regex`es are case-insensitive (except formulae in allowlist).
7. Ensure that the `regex` is enclosed within parentheses.

I haven'[t included the cop for counting number of capturing groups – I wasn't able to write a regex that was able to do this, and I couldn't find a method/function to count capturing groups without matching.

The `regex` for the extension checking needs to be verified. Here's a more convoluted version I wrote that matches all possible `tar` extensions (I think): https://rubular.com/r/d5uJPJIONM63VE

It is possible that I've missed some important points in this PR description, but I will answer any questions as they arise. Looking forward to your reviews and suggestions everyone. Thanks!

CC @samford

nandahkrishna commented:
Ah, I'd forgotten to mention: CI is 🔴 because of some recent updates to livecheck URLs in homebrew/core, which violate a cop this PR introduces. I'll create a PR to fix those issues.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/audit.rb:
```diff
@@ -891,6 +893,21 @@ def audit_revision_and_version_scheme
     end
 
+    def audit_livecheck
+      return unless formula.livecheckable?
```
Should also check that `--online` is passed.

MikeMcQuaid commented on Library/Homebrew/utils/livecheck_formula.rb:
```diff
@@ -27,4 +27,68 @@ def parse_livecheck_response(response)
     }
   end
+
+  # A class that can be used to mimic the parsed command-lione arguments used
```
```suggestion
  # A class that can be used to mimic the parsed command-line arguments used
```

MikeMcQuaid commented on Library/Homebrew/utils/livecheck_formula.rb:
```diff
@@ -27,4 +27,68 @@ def parse_livecheck_response(response)
+  # by livecheck. Necessary parameter (`args`) for `Livecheck.livecheck_formulae`.
+  #
+  # @api private
+  class LivecheckArgs
```
Feels like this should be part of `audit`?

nandahkrishna replied:
We might need to use it here as well. Currently the `LivecheckFormula` module uses the output of the `dev-cmd`, but with this class it can now call the `livecheck_formulae` method internally. That's why I moved it here (and it also seems like a 'utility' class to use anywhere we want to call `livecheck_formulae`).

If it has to be in audit, I'll move it.

MikeMcQuaid replied:
Doesn't have to move. It does make me wonder if it wouldn't be better to change the interface for `Homebrew::Livecheck.livecheck_formulae` to take the same arguments as these instead and `LivecheckFormula` as a class should go away entirely. If you're feeling even more ambitious it'd be nice to have the `livecheck`-calling logic extracted from the `Repology` module because it feels a bit weird to be there.

None of the `LivecheckFormula` or `Repology` stuff needs to (or should really) be in this PR, though!

nandahkrishna replied:
I've worked around this using `OpenStruct`: if the arguments are passed as a hash, I use `OpenStruct` to make them accessible using methods. So we don't need the `LivecheckArgs` class any more.

I've made note of the `LivecheckFormula` and `Repology` suggestions, I'll open up a PR soon after this one's done 👍🏼

nandahkrishna commented:
With the latest commit, some changes have been made:

* `livecheck_formula` now accepts a hash containing the arguments (for the `args` parameter), and this is converted into an `OpenStruct` to be able to access the arguments via methods. As a result the `LivecheckArgs` class I'd earlier added is no longer required.

* Fixed errors and incorporated review suggestions.

nandahkrishna commented:
The tests I added pass locally (`brew tests --online --only dev-cmd/audit`), but this failure seems to be related to the general issue of livecheck network tests not working on CI. If it's okay, I'll remove that particular test alone for now, and  add it once we've arrived at a solution for the failures.

SMillerDev commented on Library/Homebrew/dev-cmd/audit.rb:
```diff
@@ -902,6 +903,32 @@ def audit_revision_and_version_scheme
+      return unless formula.livecheckable?
+      return unless @online
+
+      begin
```
I think the better way is to abstract the printing out of the `livecheck_formulae` function so you can just check if any data is returned instead of jumping through a bunch of hoops like this.

In my ideal world this would just run
```ruby
problem "" if Homebrew::Livecheck.livecheck_formulae([formula]).nil
```

---

You might also just want to skip the audit for this round

nandahkrishna replied:
Since those changes should probably be part of another PR, I guess I'll remove the audit for now.

SMillerDev replied:
Sounds good to me, better to do small iterations.

MikeMcQuaid replied:
Agreed on pulling this into another PR 👍🏻

MikeMcQuaid commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -40,6 +41,9 @@ module Livecheck
     # and prints the results.
     # @return [nil]
     def livecheck_formulae(formulae_to_check, args)
+      # If args is a hash, convert it to an OpenStruct so that options can be accessed using methods
```
Would be nicer to use keyword arguments here as otherwise this'll break on Ruby 2.7.

nandahkrishna replied:
Would this be okay?
```ruby
def livecheck_formulae(formulae_to_check, args:)
```
`args` isn't always going to be a hash, it could even be parsed arguments from the parser. Calling the method using `livecheck_formulae(formulae_to_check, args: args)` where `args` is either a hash or the parsed args would prevent the warning/incompatibility, wouldn't it?

Wasn't sure if you meant this or something like `def livecheck_formulae(formulae_to_check, quiet: false, verbose: false, ...)`. Let me know what you think!

MikeMcQuaid replied:
> Wasn't sure if you meant this or something like `def livecheck_formulae(formulae_to_check, quiet: false, verbose: false, ...)`

Yeh, this ideally 👍🏻

nandahkrishna replied:
I'm moving this suggestion to another PR that addresses the suggestion on abstracting the `puts` out of `livecheck_formulae` and the audit, as it's not required in this PR.

MikeMcQuaid commented on Library/Homebrew/rubocops/livecheck.rb:
```diff
@@ -0,0 +1,245 @@
+          return if livecheck_node.blank?
+
+          skip = find_every_method_call_by_name(livecheck_node, :skip).first
+          return unless skip.present? && find_every_method_call_by_name(livecheck_node).length > 2
```
```suggestion
          return if skip.blank? && find_every_method_call_by_name(livecheck_node).length < 3
```

MikeMcQuaid commented on Library/Homebrew/rubocops/livecheck.rb:
```diff
@@ -0,0 +1,245 @@
+            skip = string_content(skip) if skip.present?
+            corrector.replace(
+              node.source_range,
+              "livecheck do\n    skip#{" \"#{skip}\"" if skip.present?}\n  end",
```
Use `EOS` here rather than embedded newlines.

MikeMcQuaid commented on Library/Homebrew/rubocops/livecheck.rb:
```diff
@@ -0,0 +1,245 @@
+          formula_urls = { head: head_url, stable: stable_url, homepage: homepage_url }.compact
+
+          formula_urls.each do |symbol, url|
+            next unless url == livecheck_url || url == "#{livecheck_url}/" || "#{url}/" == livecheck_url
```
Can you make this an `if` instead?

MikeMcQuaid reviewed (approved):
A few nits but otherwise looks great, nice work!

nandahkrishna commented:
I've made all the changes @MikeMcQuaid, let me know if this is good to merge. I'll create a separate PR to address the livecheck `puts` suggestion and audit.
